### DISCUSSION TITLE
Model Evaluation with backtesting_sarimax

### DISCUSSION THREAD

**Discussion by samuel-gerstein** [JoaquinAmatRodrigo/skforecast#600] [Q&A] [answered]
I know with backtesting_forecaster you can evaluate the entire dataset and get the predictions for the train set by setting initial_train_size = None. Is there a similar way to do this with backtesting_sarimax? If not, how can I make predictions on the train set?

**Comment by JavierEscobarOrtiz** [comment]
Hello @samuel-gerstein,

The predictions of the fitted values are stored in the `fittedvalues` attribute of the `SARIMAXResults` object:

https://www.statsmodels.org/dev/generated/statsmodels.tsa.statespace.sarimax.SARIMAXResults.fittedvalues.html#statsmodels.tsa.statespace.sarimax.SARIMAXResults.fittedvalues

This may be a bit confusing, but the code is:

```python
# Create and fit ForecasterSarimax
# ==============================================================================
forecaster = ForecasterSarimax(
                 regressor=Sarimax(order=(12, 1, 1), seasonal_order=(0, 0, 0, 0), maxiter=200),
             )

forecaster.fit(y=data_train['y'], suppress_warnings=True)

# In-sample Predictions
# ==============================================================================
forecaster.regressor.sarimax_res.fittedvalues
```

This will return a pandas Series with the predictions of the fitted values.

The statsmodels `SARIMAXResults` object is stored inside the `sarimax_res` attribute of the `skforecast.Sarimax` model:

https://skforecast.org/latest/api/sarimax

Thank you for pointing this out. We will include it in the user guide. 😄

**Reply by JavierEscobarOrtiz** [reply]
Here is: https://skforecast.org/latest/user_guides/forecasting-sarimax-arima#prediction-on-training-data-in-sample-predictions

**Reply by samuel-gerstein** [reply]
Thanks for your prompt reply! Quick clarification, is there a way to capture those fitted values within backtesting_sarimax for an n steps ahead forecast? I wanted to replicate the process done in [] (https://skforecast.org/0.11.0/user_guides/backtesting#backtesting-on-training-data). 

I'm currently training SARIMAX on a next hour and next 24hr forecast and wanted to compare predictions on the entire dataset.

**Reply by JavierEscobarOrtiz** [reply]
Regarding the backtesting-on-training-data, you cannot do this with the backtesting_sarimax. The reason is that the last window of the Sarimax model doesn't work in the same way as, for example, the one in the ForecasterAutoreg. 

In the Sarimax models, the `last_window` (values needed to predict the next step) must start at the end of the training data. For example, you cannot train with 100 observations and then use observations 16 to 20 (5 lags) as `last_window` because the forecaster expects observations 101 to 105.

So, for your use case, I will recommend using a [backtesting with refit](https://skforecast.org/latest/user_guides/backtesting#backtesting-with-refit-and-increasing-training-size-fixed-origin) strategy. You can start with a small green area (`initial_train_size`) and set `refit = True` and `fixed_train_size = False`. It is true that you will lose the first `initial_train_size` observations, but this is the fairest way to compare two models, since you are comparing all time test values.

<a href="https://skforecast.org/0.11.0/img/backtesting_refit.gif" target="_blank">
  <img src="https://skforecast.org/0.11.0/img/backtesting_refit.gif" alt="Backtesting refit" width="550"/>
</a>

**Reply by samuel-gerstein** [reply]
Is it possible to modify the training process then to calculate the fittedvalues for n-step predictions? It seems like this can be done quite simply in R, see fig 12.8 [here](https://otexts.com/fpp2/forecasting-on-training-and-test-sets.html) and the fitted function [here](https://rdrr.io/cran/forecast/man/fitted.Arima.html).

**Reply by JavierEscobarOrtiz** [reply]
Right now it is not possible using `backtesting_sarimax`. I think this may be possible using the statsmodels `SARIMAXResults` object inside the `sarimax_res` attribute of the skforecast.Sarimax model and its `apply` method:

https://www.statsmodels.org/dev/generated/statsmodels.tsa.statespace.mlemodel.MLEResults.apply.html#statsmodels.tsa.statespace.mlemodel.MLEResults.apply

**Reply by samuel-gerstein** [reply]
> Right now it is not possible using `backtesting_sarimax`. I think this may be possible using the statsmodels `SARIMAXResults` object inside the `sarimax_res` attribute of the skforecast.Sarimax model and its `apply` method:
> 
> 
> 
> https://www.statsmodels.org/dev/generated/statsmodels.tsa.statespace.mlemodel.MLEResults.apply.html#statsmodels.tsa.statespace.mlemodel.MLEResults.apply

That makes sense. Wouldn't it just be as simple as accessing `sarimax_res`from the copy of forecaster in `backtesting_sarimax`?

**Reply by JavierEscobarOrtiz** [reply, marked as answer]
Yes, you are right, that makes sense. Will look into it for the next release. 😄
